polyfill/lib/calendar.mjs

```diff
@@ -333,53 +332,1354 @@ function resolveNonLunisolarMonth(calendarDate) {
+          next = this.addMonthsCalendar(current, sign, 'constrain', cache);
+          if (next.day !== calendarOne.day) {
+            // In case the day was constrained down, try to un-constrain it
+            next = this.regulateDate({ ...next, day: calendarOne.day }, 'constrain', cache);
+          }
+        } while (this.compareCalendarDates(calendarTwo, next) * sign >= 0);
+        months--; // correct for loop above which overshoots by 1
+        const remainingDays = this.calendarDaysUntil(current, calendarTwo, cache);
+        days = remainingDays % 7;
+        weeks = (remainingDays - days) / 7;
+        break;
+      }
     }
+    return { years, months, weeks, days };
+  },
+  daysInMonth(calendarDate, cache) {
+    // Add the length of the smallest possible month. If it rolls over to the
+    // next month, then back up to the end of the previous month to know its day
+    // count. If it doesn't roll over, add the same number of days again until
+    // we get a rollover.
+    //
+    // NOTE: This won't work for months where days are skipped (e.g. Gregorian
+    // switchover) which is OK because this is a prototype implementation.
+    const { month: originalCalendarMonth } = calendarDate;
+    const increment = this.minimumMonthLength(calendarDate);
+
+    // easiest case: we already know the month length if min and max are the same.
+    if (increment === this.maximumMonthLength(calendarDate)) return increment;
+
+    let addedIsoDate = this.calendarToIsoDate(calendarDate, 'constrain', cache);
+    let addedCalendarDate;
+    do {
+      addedIsoDate = this.addDaysIso(addedIsoDate, increment);
+      addedCalendarDate = this.isoToCalendarDate(addedIsoDate, cache);
+    } while (addedCalendarDate.month === originalCalendarMonth);
+
+    // Now back up to the last day of the original month
+    const endOfMonthIso = this.addDaysIso(addedIsoDate, -addedCalendarDate.day);
+    const endOfMonthCalendar = this.isoToCalendarDate(endOfMonthIso, cache);
+    return endOfMonthCalendar.day;
+  },
+  daysInPreviousMonth(calendarDate, cache) {
+    const { day, month, year } = calendarDate;
+
+    // Check to see if we already know the month length, and return it if so
+    const previousMonthYear = month > 1 ? year : year - 1;
+    let previousMonthDate = { year: previousMonthYear, month, day: 1 };
+    const previousMonth = month > 1 ? month - 1 : this.monthsInYear(previousMonthDate, cache);
+    previousMonthDate = { ...previousMonthDate, month: previousMonth };
+    const min = this.minimumMonthLength(previousMonthDate);
+    const max = this.maximumMonthLength(previousMonthDate);
+    if (min === max) return max;
+
+    const isoDate = this.calendarToIsoDate(calendarDate, 'constrain', cache);
+    const lastDayOfPreviousMonthIso = this.addDaysIso(isoDate, -day);
+    const lastDayOfPreviousMonthCalendar = this.isoToCalendarDate(lastDayOfPreviousMonthIso, cache);
+    return lastDayOfPreviousMonthCalendar.day;
```

Comment:
Would this approach (or directly calling this) not work for daysInMonth? It would involve no loop and fewer caching calls. Not that the loop executes much at all, I hope, for any ICU calendars.

Reply:
> Not that the loop executes much at all, I hope, for any ICU calendars.

That's correct. In an earlier version of this PR, it was worse before because the minimum month length for `ethiopic/`coptic`/`ethioaa` is 5 days, which meant a lot of looping. But then I changed things to calculate the min/max month lengths based on calendar and the entire calendar date, so now the maximum number of loop iterations for any ICU calendar is going to be 2.

> Would this approach (or directly calling this) not work for daysInMonth? It would involve no loop and fewer caching calls.

Good suggestion! The reason the loop is needed is because (unlike when measuring the previous month) it's not as easy to deduce the correct number of days to add to guarantee that you can get to the next month. For the previous month, you know you always need to subtract `day` to get to the last day of the previous month, and the `day` of that addition result is the number of days in that previous month. However, for the current month forward, that trick doesn't work because you don't know the month length so don't know exactly how many days to guarantee that the result is in the next month. If you add the minimum month length, then if `day===1` then you might be stuck in the same month. But if you add the maximum month length, then you might skip the next month entirely if you started on the last day of the current month.

But your comment made me realize that you're correct there's an O(1) solution here, which is to add a variable number of days depending on `day`.  If `day <= max-min` then you can safely add the maximum month length because you know you won't overflow. Otherwise, you can add the minimum length because you know you're guaranteed to end up in the next month.

Ordinarily I'd say "don't bother" for this small optimization, but "first day of the month" is probably the most common use case and this further optimization could potentially speed up first-day-of-month calls to this method by 30%-40% so seems worth doing. I'll try it and if there's a significant perf benefit, I'll leave it in.

Thanks for catching this!

Reply:
I tried this change and it does seem to have a small impact on test runtime, but not much, maybe 5% faster at best (lots of noise in these perf tests so hard to know for sure). But the code is simpler without the loop, so I'll keep it!

Reply:
Awesome! Yes, my comment was also thinking that it seemed simpler and there was no reason I could think of a similarly constant approach wouldn't work for the other method - but the devil's in the details, thanks for figuring them out :)